runtime/src/main/cpp/Memory.cpp

```diff
@@ -2708,6 +2808,118 @@ void shareAny(ObjHeader* obj) {
   container->makeShared();
 }
 
+ScopedRefHolder::ScopedRefHolder(KRef obj): obj_(obj) {
+  if (obj_) {
+    addHeapRef(obj_);
+  }
+}
+
+ScopedRefHolder::~ScopedRefHolder() {
+  if (obj_) {
+    ReleaseHeapRef(obj_);
+  }
+}
+
+// static
+CycleDetectorRootset CycleDetector::collectRootset() {
+  auto& detector = instance();
+  CycleDetectorRootset rootset;
+  LockGuard<SimpleMutex> guard(detector.lock_);
+  for (auto* candidate: detector.candidateList_) {
+    rootset.roots.push_back(candidate);
+    rootset.heldRefs.emplace_back(candidate);
+    traverseReferredObjects(candidate, [&rootset, candidate](KRef field) {
+      rootset.rootToFields[candidate].push_back(field);
+      rootset.heldRefs.emplace_back(field);
```

Comment:
I suspect a data race here: this thread reads field, other thread nullifies the field and performs GC, which removes the original field value.
We have a lock in atomic reference to protect from such a race. Using it here would probably be an overkill, but we should at least add a comment about this.

Reply:
Okay, so I left a comment for now, but I'm very unhappy about this overlooked data race.

Reply:
I guess it is ok for this cycle detector tool.